### PR TITLE
ENH: stats.anderson_ksamp: add permutation version of test

### LINKED ISSUE #9527: Anderson-Darling ksamples can not estimate p-values beyond given range
I've been following the trace for this long standing issue. let me do a quick recap:

It was apparently first reported in 2015 (https://github.com/scipy/scipy/issues/5493) mentioning that the test was throwing bad p-values (larger than 1). A similar issue was reported this year (https://github.com/scipy/scipy/issues/8536) with the discussion apparently agreeing on capping the p-values at 0.25. This was addressed by this commit https://github.com/scipy/scipy/pull/8560.

Both issues above are now closed, but I think the problem remains. Capping the p-values at 0.25 sort of works, but it does not "fix" the test. My main question is (this also came up in the threads mentioned above): why can't we do whatever is done by the [kSamples](https://cran.r-project.org/web/packages/kSamples/index.html) R package?

I did some tests with the `ad.test` function in that package, and it does not have the problems with the p-values that [anderson_ksamp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.anderson_ksamp.html) has.

Can't whatever is done in `ad.test` be incorporated into `anderson_ksamp`?

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-9527

#### What does this implement/fix?
gh-9527 asked for the `anderson_ksamp` to provide p-values beyond the range of tabulated values. This PR does that by performing a permutation test rather than relying on interpolation from tables.

```python3
import numpy as np
from scipy import stats
rng = np.random.default_rng(24985724)
samples = rng.random(size=(3, 20))
samples[0] += 0.15
ref = stats.anderson_ksamp(samples)
res = stats.anderson_ksamp(samples, n_resamples=9999, random_state=rng)
print(res.pvalue, ref.pvalue)  # 0.03 0.030398697650025144

samples[0] -= 0.15
ref = stats.anderson_ksamp(samples)
# UserWarning: p-value capped: true value larger than 0.25
res = stats.anderson_ksamp(samples, n_resamples=9999, random_state=rng)
print(res.pvalue, ref.pvalue)  # 0.6698 0.25 
```

### CODE REVIEW

File: scipy/stats/_morestats.py [hunk 1]
```diff
@@ -2143,7 +2158,7 @@ def anderson_ksamp(samples, midrank=True):
     --------
     >>> import numpy as np
     >>> from scipy import stats
-    >>> rng = np.random.default_rng()
+    >>> rng = np.random.default_rng(1638083107694713882823079058616272161)
```
Comment (mdhaber) [inline scipy/stats/_morestats.py:2161]:
```suggestion
    >>> rng = np.random.default_rng()
```

File: scipy/stats/tests/test_morestats.py [hunk 2]
```diff
@@ -359,6 +359,7 @@ def test_example1b(self):
                                   tm[0:5], 4)
         assert_allclose(p, 0.0020, atol=0.00025)
 
+    @pytest.mark.slow
```
Comment (tupui) [inline scipy/stats/tests/test_morestats.py:362]: How slow is this?

**Review by tupui — COMMENTED** [review]
Thanks @mdhaber, a few suggestions.

File: scipy/stats/_morestats.py [hunk 3]
```diff
@@ -2085,6 +2086,20 @@ def anderson_ksamp(samples, midrank=True):
         (True) is the midrank test applicable to continuous and
         discrete populations. If False, the right side empirical
         distribution is used.
+    n_resamples : int, default: 0
+        If positive, perform a permutation test to determine the p-value
+        rather than interpolating between tabulated values.
+    random_state : {None, int, `numpy.random.Generator`,
+                    `numpy.random.RandomState`}, optional
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/_morestats.py:2093]: Ok. Yeah at some point maybe we should do it across the board. Or doc-deprecate `RandomState`. You'd love that!

File: scipy/stats/_morestats.py [hunk 4]
```diff
@@ -2067,7 +2068,7 @@ def _anderson_ksamp_right(samples, Z, Zstar, k, n, N):
 )
 
 
-def anderson_ksamp(samples, midrank=True):
+def anderson_ksamp(samples, midrank=True, n_resamples=0, random_state=None):
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/_morestats.py:2071]: Yup good idea.

File: scipy/stats/tests/test_morestats.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/tests/test_morestats.py:362]: 2.71s.

**Comment by mdhaber** [comment]
All CirrusCI failures seem unrelated.

**Comment by andyfaff** [comment]
Don't know why cirrus is failing on this PR and not on others.

File: scipy/stats/_morestats.py [hunk 5]
```diff
@@ -2236,19 +2268,23 @@ def anderson_ksamp(samples, midrank=True):
...
 
     sig = np.array([0.25, 0.1, 0.05, 0.025, 0.01, 0.005, 0.001])
-    if A2 < critical.min():
+    if A2 < critical.min() and not n_resamples:
         p = sig.max()
-        warnings.warn("p-value capped: true value larger than {}".format(p),
-                      stacklevel=2)
-    elif A2 > critical.max():
+        message = (f"p-value capped: true value larger than {p}. Consider "
+                   "setting `n_resamples` to a possible integer (e.g. 9999).")
```
Comment (tupui) [inline scipy/stats/_morestats.py:2274]: I think you meant "positive"?

**Review by tupui — APPROVED** [review]
LGTM, one last question before merging. (I re-launch Cirrus' CI just to see, but not linked to this PR anyway.)

**Comment by tupui** [comment]
Cirrus failed again. Is this branch up to date with master? Maybe there was a submodule update and this branch is off and Cirrus does not properly do the merge before running CI (like Circle)

File: scipy/stats/_morestats.py (hunk 5 above)
Comment (mdhaber) [reply scipy/stats/_morestats.py:2274]:
```suggestion
                   "setting `n_resamples` to a positive integer (e.g. 9999).")
```

**Comment by mdhaber** [comment]
The branch is up to date. I merged main and it doesn't look like there are any submodules out of date. Weird!

**Comment by tupui** [comment]
CI is getting back green, in it goes. Thanks again Matt!

**Comment by mdhaber** [comment]
@tupui I've considered doing this sort of thing for essentially all stats functions. For instance, gh-17530 just brought up again that `wilcoxon` can't compute exact p-values when there are ties or zeros, and the same is true of `mannwhitneyu`. For many other stats functions (e.g. t-tests) there is a corresponding permutation test that can give exact p-values. 

Does this sound good to you? Do we mind adding two separate parameters to these functions `random_state` and `n_resamples`, or can we think of a smaller change to the API? 

For instance, we could create a class or dictionary that contains information about the number of resamples and the random state, and this could be passed into `wilcoxon`'s `method` parameter, which seems reasonable because it's specifying that the method should be done via permutations rather than either the "exact" or "approximate" methods. I'm not saying this is a good option; it's just an example of a way of providing this information in a way that would not complicate the signatures.

**Comment by tupui** [comment]
The change seems sensible to me. In general I am not a fan of passing dicts as it makes it harder when coding to know what can be in the dict. For a user it is faster/simpler to play directly with parameters. Happy to discuss this further.

**Comment by mdhaber** [comment]
I agree about dicts. I actually would have preferred that `minimize` and `linprog` would accept options as keyword arguments but just document them in a separate options section. (I suppose we could still do that, actually. It could be done in a BC way simply by accepting `**kwargs` and doing `options.update(kwargs)`, I think, so that information from both is accepted.)

Another option I had in mind was that the user could pass in a `callable` rather than a random state, which we could call repeatedly for random permutations until it raises a `StopIteration`. That way, the user could control the batch size, too. The intent is to not have to recreate the interface of `permutation_test` in every function. I want to avoid the `minimizer_kwargs` idea of global optimizers.

**Pull request merged**